components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java

```diff
@@ -431,6 +451,75 @@ public void deleteBotDataEmailList(String uuid) throws APIManagementException, S
         apiMgtDAO.deleteBotDataEmailList(uuid);
     }
 
+    /**
+     * Retrieve all bot detected data
+     *
+     * @return list of bot detected data
+     * @throws APIManagementException
+     */
+    public List<BotDetectionData> retrieveBotDetectionData() throws APIManagementException {
+
+        List<BotDetectionData> botDetectionDatalist = new ArrayList<>();
+        String appName = AlertMgtConstants.APIM_ALERT_BOT_DETECTION_APP;
+        String query = SQLConstants.BotDataConstants.GET_BOT_DETECTED_DATA;
+
+        JSONObject botDataJsonObject = APIUtil.executeQueryOnStreamProcessor(appName, query);
+        if (botDataJsonObject != null) {
+            JSONArray botDataJsonArray = (JSONArray) botDataJsonObject.get("records");
+            if (botDataJsonArray != null && botDataJsonArray.size() != 0) {
+                for (Object botData : botDataJsonArray) {
+                    JSONArray values = (JSONArray) botData;
+                    BotDetectionData botDetectionData = new BotDetectionData();
+                    botDetectionData.setCurrentTime((Long) values.get(0));
+                    botDetectionData.setMessageID((String) values.get(1));
+                    botDetectionData.setApiMethod((String) values.get(2));
+                    botDetectionData.setHeaderSet((String) values.get(3));
+                    botDetectionData.setMessageBody(extractBotDetectionDataContent((String) values.get(4)));
+                    botDetectionData.setClientIp((String) values.get(5));
+                    botDetectionDatalist.add(botDetectionData);
+                }
+            }
+        }
+        return botDetectionDatalist;
+    }
+
+    /**
+     * Extract content of the bot detection data
+     *
+     * @param messageBody message body of bot detection data
+     * @return extracted content
+     */
+    public String extractBotDetectionDataContent(String messageBody) {
+
+        String content = "";
+        try {
+            //Parser that produces DOM object trees from XML content
+            DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
```

Comment:
Shall we use APIUtil.getSecuredDocumentBuilder() because, to avoid XXE attacks. (If the bot payload contains any entity tags, the secured doc builder avoids processing them.

Reply:
Fixed with https://github.com/wso2/carbon-apimgt/pull/8708/commits/ffc1f632169b3861b7c65b4e05f2ffa0f8d71ee9